### DISCUSSION TITLE
knitr::read_chunks and knitr::purl for other languages

### DISCUSSION THREAD

**Discussion by baptiste** [quarto-dev/quarto-cli#9112] [Feature Requests]
### Description

I'm a big fan of knitr's code externalisation: it lets me tinker with the code in a self-contained document, reuse it in other files (e.g. `source('funs.R')`) etc. while focusing more on the text/narrative in the Rmd/qmd file, but still benefit from the reproducibility of literate programming (namely: the code is the actual code, and all artefacts have resulted from it).

Unfortunately, I believe this only exists for the knitr+R combination, not for other languages or engines?

As a workaround, I've resorted to a reverse strategy, where all the code is in a qmd file, and I use `knitr::purl` to extract it to a standalone source file.

````
---
title: "testpurl"
engine: knitr
---

```{julia, eval=FALSE}
1 + 1
```


```{julia, eval=FALSE}
plot(1,1)
```

`r invisible(knitr::purl(xfun::with_ext(knitr::current_input(), "qmd"),output=xfun::with_ext(knitr::current_input(), "jl")))`
````

Note the R code at the end is not in a chunk, lest it be included in the julia script (and would cause errors). It's a bit awkward, I think it'd be nice if quarto had such an option to "dump" all code into a script at the user's request? (and ideally this script could be listed as "other formats" in a quarto webpage). (So I guess this is 3 feature requests into one)

**Comment by allefeld** [comment]
I would like to have this feature, too.

My use-case are `qmd` files published as a website, including Matlab code executed via a custom Jupyter kernel. I didn't think to use `knitr::purl`, but `engine: knitr` is not an option for me anyway.

What I ended up doing is to implement a conversion from Jupyter notebook to Matlab script in Matlab, called from a code cell with `echo: false`. That notebook exists (with extension `.quarto_ipynb`) during execution, because that is the way Quarto executes Jupyter-kernel based `qmd` files. Implementation is relatively easy, Jupyter notebooks are quite simple JSON structures.

But such a conversion could be implemented in Quarto itself, activated e.g. via a YAML header
```yaml
format:
  html: default
  script: default
```
and offered for download as an ["other format"](https://quarto.org/docs/output-formats/html-multi-format.html).

Somewhat related: #7411.

**Reply by cscheid** [reply]
This is the kind of feature that should be implemented by Jupyter kernels and languages themselves; there's no way that Quarto can design a solution that works for all engines and programming languages with their myriad semantics and way to express programs.

**Reply by baptiste** [reply]
I must be missing something, because in my mind something like `quarto purl mydoc.qmd` could ask pandoc to look for AST nodes corresponding to code chunks and just dump them all into a text file? (and potentially some optional arguments could filter e.g. language types, chunk options, etc.)

**Reply by cscheid** [reply]
... wait - I think I confused `purl` with other literate programming features! If "all" that's needed is to collect all code cells, then you can probably do that in 1.6 with `quarto inspect` and an iteration over the resulting JSON:

## `purl-test.qmd`

````qmd
## Section

```{r}
cat("Hello, world")
```

## Another section

```{r}
cat("more")
```
````

## shell

```
$ quarto inspect purl-test.qmd | jq '.fileInformation["purl-test.qmd"].codeCells'
[
  {
    "start": 2,
    "end": 4,
    "file": "purl-test.qmd",
    "source": "cat(\"Hello, world\")\n",
    "language": "r",
    "metadata": {}
  },
  {
    "start": 8,
    "end": 10,
    "file": "purl-test.qmd",
    "source": "cat(\"more\")\n",
    "language": "r",
    "metadata": {}
  }
]
```

Notably, `quarto inspect` knows to resolve include shortcodes correctly.

**Reply by cscheid** [reply]
Or, more directly (and assuming it's all one language):

```
$ quarto inspect purl-test.qmd | jq -r '[.fileInformation["purl-test.qmd"].codeCells[].source] | add'
cat("Hello, world")
cat("more")
```

**Reply by cscheid** [reply]
Note: this will miss inline code cells, get things wrong with `eval: false` (to say nothing of evil stuff like `eval: !expr ...`), etc. But I think that's what you were asking for.

**Reply by baptiste** [reply]
Nice! Yes, that's basically what's needed for the purl() side. It would be nice I think to have a user-friendly wrapper (maybe `quarto convert` or a new name if that's too different because lossy conversion) that takes care of the post-processing the json and spits out a runnable script (assuming one language, but it makes little sense otherwise).

The other side, `knitr::read_chunk()`, I'm not sure how to approach. I really like code externalisation in knitr, and I imagine a similar feature could be generalised (it's just reading custom comments and turning them into chunk options, which is trivial now that chunk options are yaml comments anyway even on the qmd side), but I don't know how quarto would load and refer to external code in engines beside knitr.

**Reply by mcanouil** [reply]
`quarto convert` is already used for Quarto <=> Jupyter.

Note that you can "easily" extract multiple languages blocks with this approach.

````qmd
## Section

```{r}
cat("Hello, world")
```

## Another section

```{r}
cat("more")
```

## Python

```{python}
print("Hello, world")
```
````

```sh
quarto inspect index.qmd | jq '
  .fileInformation[].codeCells |
  group_by(.language) |
  map({language: .[0].language, sources: map(.source) | join("\n")})
'
```
```json
[
  {
    "language": "python",
    "sources": "print(\"Hello, world\")\n"
  },
  {
    "language": "r",
    "sources": "cat(\"Hello, world\")\n\ncat(\"more\")\n"
  }
]
```

**Reply by mcanouil** [reply]
You could even write the metadata such as label.
This means you could turn Quarto document as script files you can render with Quarto (https://quarto.org/docs/computations/render-scripts.html).

<details><summary>Quarto document</summary>

````qmd
## Section

```{r}
#| label: r-cell-1
#| echo: true
#| eval: false
cat("Hello, world")
```

## Another section

```{r}
#| label: r-cell-2
cat("more")
```

## Python

```{python}
#| label: python-cell-1
print("Hello, world")
```
````

</details>

```sh
quarto inspect index.qmd | jq '
  .fileInformation[].codeCells |
  map(
    .metadata = (.metadata | to_entries | map("#| \(.key): \(.value)") | join("\n"))
  ) |
  group_by(.language) |
  map({
    language: .[0].language,
    contents: map(.metadata + "\n" + .source) | join("\n")
  })
'
```

```json
[
  {
    "language": "python",
    "contents": "#| label: python-cell-1\nprint(\"Hello, world\")\n"
  },
  {
    "language": "r",
    "contents": "#| label: r-cell-1\n#| echo: true\n#| eval: false\ncat(\"Hello, world\")\n\n#| label: r-cell-2\ncat(\"more\")\n"
  }
]
```

Version that writes code to files (bash).

```sh
quarto inspect index.qmd | jq '
  .fileInformation[].codeCells |
  map(
    .metadata = (.metadata | to_entries | map("#| \(.key): \(.value)") | join("\n"))
  ) |
  group_by(.language) |
  map({
    language: .[0].language,
    contents: map("\n\n# %%\n" + .metadata + "\n" + .source) | join("\n")
  })
' | jq -c '.[]' | while read -r cell; do
  language=$(printf "%s\n" "$cell" | jq -r '.language')
  contents=$(printf "%s\n" "$cell" | jq -r '.contents')
  case "$language" in
    python)
      echo "$contents" > script.py
      ;;
    r)
      echo "$contents" > script.r
      ;;
    *)
      echo "Unsupported language: $language"
      ;;
  esac
done
```

<table>
<tr><th><code>script.r</code></th><th><code>script.py</code></th></tr>
<tr><td>

````r


# %%
#| label: r-cell-1
#| echo: true
#| eval: false
cat("Hello, world")



# %%
#| label: r-cell-2
cat("more")

````

</td><td>

````python


# %%
#| label: python-cell-1
print("Hello, world")

````

</td></tr>
</table>

**Reply by baptiste** [reply]
Indeed, that's the sort of thing I'd love to see built-in, as it's a common request to get a standalone script version of the code without the surrounding text.

**Reply by allefeld** [reply]
@baptiste, if it's just about the conversion from `qmd` to a script, that's not too hard.  I wrote a bit about existing tools [here](https://github.com/quarto-dev/quarto-cli/discussions/7683#discussioncomment-10369203). There is `nbconvert`, but much better is `jupytext` which now understands the `qmd` format and supports 32 different programming languages.

For me, __the reason it would be neat to have this built into quarto__ as a `format` is that then the script could be included in HTML output as an ["other format"](https://quarto.org/docs/output-formats/html-multi-format.html), see my #7411.

Such a feature could rely on an external tool, preferably `jupytext`, which has to be separately installed.

(An alternative proposed to me by @mcanouil would be to use `code-tools` / `source` for this purpose, but preferably only after it is integrated with "Other formats", see his #7436.)

***

But also @cscheid, I don't think it would be that complicated to implement the conversion in Quarto itself – but maybe I'm missing something? For knitr it can be based on `purl`, and for Jupyter all the code  and all the Markdown text can be accessed from the JSON data in the `.quarto-ipynb` file which is created anyway. There are three things this functionality needs, and they could be provided as options to the `script` format.

```yaml
format:
  script:
    language: python  # code blocks of which language should be included?
    extension: py     # what should the file extension for the script be?
    comment: "#"      # which comment character should be used?
```

This way Quarto doesn't need to know anything about specific languages.

(As far as I can see, a script conversion which includes code blocks of more than one language doesn't make sense.)

As for `quarto inspect`, it provides a lot of useful information, but it does not include the Markdown part which imho should be included in the script as comments.

**Reply by cscheid** [reply]
Those are the options _you_ want. But if we're going to do this as a feature, Quarto needs to define and implement the options that _anyone_ could possibly want, and I don't currently believe we can do this better than teaching people about `quarto inspect` so they can write their own downstream tooling.

**Reply by allefeld** [reply]
I don't understand. Of course many more features are conceivable and may be nice, but those are the bare-minimum options to create a script file which includes Markdown text as comments. And that seems to be what people typically want, judging by what `nbconvert` and `jupytext` do by default.

Anyway, what about the idea to use `jupytext` as an external tool? It's very good at figuring out the language, and knows extension and comment style for many languages. And there could be a one-size-fits-all option for `jupytext` command line flags, externalizing all the complexity (and follow-up feature requests) there.

> teaching people about quarto inspect so they can write their own downstream tooling

That's definitely a good idea – I only found about it recently and wish I had earlier.
(Maybe a page documenting all Quarto CLI commands would be good? `quarto --help` is a bit sparse.)
But I still think for conversion the JSON-format `ipynb` is more useful.

**Reply by mcanouil** [reply]
> Maybe a page documenting all Quarto CLI commands would be good? quarto --help is a bit sparse.

There is an open issue about that.

**Reply by cderv** [reply]
Nice thread and ideas ! 

FWIW I opened a FR in **quarto** R package for a wrapper function to extract script
- https://github.com/quarto-dev/quarto-r/issues/208

I will be for R user main, but can help prototype and be useful while we think if / how / when this could be a `script` format. 

Personally, I think more of this as a `quarto convert` feature than a rendered output format

**Comment by mcanouil** [comment]
For reference:

- #389

**Comment by cderv** [comment]
**quarto** R package will now have an experimental `qmd_to_r_script()` function to do like `purl()` and extract R cells in an R script. 

This is a wrapper around `quarto inspect` to get the cell information. Please do give it a try. 

As of 07/2025, this is in dev version at https://github.com/quarto-dev/quarto-r

It can be installed easily using Github or r-universe build: https://quarto-dev.r-universe.dev/quarto

See vignette: https://quarto-dev.github.io/quarto-r/articles/r-scripts.html

**Reply by srvanderplas** [reply]
It would also be good to support the purl=F chunk option (or an equivalent) to exclude the chunk from the R script -- for instance, in my use case, to exclude the timer code used in my slides from the R code files for students.

**Reply by cderv** [reply]
Can you open an issue in quarto-dev/quarto-r with an example? 

I did not try to get a 1-1 feature parity with purl, but I surely may have missed some use case.
Release is going out this week, but if this is easy enough I can maybe put it in. 

A detailed example of expected output would help me understand. thank you!

**Reply by srvanderplas** [reply]
https://github.com/quarto-dev/quarto-r/issues/277

**Reply by cderv** [reply]
Thank you

**Comment by mcanouil** [comment]
FYI

- https://github.com/quarto-dev/quarto-cli/issues/1156

**Comment by coatless** [comment]
You may want to use a new extension I created called **Ripper**, which automatically extracts code blocks from your documents and splits them by programming language into separate, executable script files.

- Demo: https://quarto.thecoatlessprofessor.com/ripper/ 
- Repo: https://github.com/coatless-quarto/ripper 

This has a bit more support for other languages compared to `qmd_to_r_script()`, which restrict the code to only _R_ as the output. This comes at the cost of not knowing about any _suppressed_ code cells (e.g. `echo: false`/`include: false`).
